### PR TITLE
session 02

### PR DESCRIPTION
**Pull request by guillaumethomas**
These are the first 2 exercises.
Second pull request mostly to test the workflow and get your feedback

thank you 

G.

### CODE REVIEW

File: students/guillaume/session02/FizzBuzz.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+			print('FizzBuzz')
+		elif i % 3 == 0:
+			print('Fizz')
+		elif i % 5 == 0:
+			print('Buzz')
+		else:
+			print(i)
+
+if __name__ == '__main__':
+	FizzBuzz()
```
Comment (codedragon) [inline students/guillaume/session02/FizzBuzz.py]: Cool!

**Review by codedragon — COMMENTED** [review]
Nice work. I'm going to wait to review this until we have worked on this stuff in class.

File: students/guillaume/session03/list_lab.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python3
```
Comment (PythonCHB) [inline students/guillaume/session03/list_lab.py:1]: you are getting quite ahead of yourself (or the rest of us, anyway) nothing wrong with that, but as this is a session 3 exercise -- we will review it then. (any maybe put it in a "session 3" dir to keep things straight.

I do apologize -- the first few sessions are going to feel slow for folks that have some Python experience. But not everyone is the class does have much experience with python, so we spend the first few classes on the basics to get everyone on the same page.

**Comment by PythonCHB** [comment]
You may want to take a breather and wait for the rest of the class to catch up!

File: students/guillaume/session02/grid.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+def example_grid():
+    '''
+    Print a simple grid
+    '''
+
+    for _ in range(2):
```
Comment (PythonCHB) [inline students/guillaume/session02/grid.py:7]: nice use of the underscore.

File: students/guillaume/session02/grid.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            print('|', 4 * ' ', '|', 4 * ' ', '|')
+
+    print('+', 4 * '-', '+', 4 * '-', '+')
+
+
+def print_grid(n):
+    '''
+    n is the length of the line minus the plus on the side
+    '''
+    if n < 3:
```
Comment (PythonCHB) [inline students/guillaume/session02/grid.py:20]: you could also do this with:

n = max(n, 3)

File: students/guillaume/session02/grid.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        lines[i] = line_a
+
+    for line in lines:
+        print(line)
+
+
+def gene_line(char_a, char_b, n):
+
+    line = char_a + ' ' + ((n - 1) // 2) * (char_b + ' ')
+    line = line + char_a + line[::-1]
```
Comment (PythonCHB) [inline students/guillaume/session02/grid.py:55]: you can use +=

line += char_a + line[::-1]

And very nice use of slicing to reverse the string!

File: students/guillaume/session02/grid.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+def gene_line2(char_a, char_b, m, n):
+
+    sub_line =  char_a + ' ' + n * ( char_b + ' ' )
+    line = sub_line * m + char_a
+    return line
+
+
+if __name__ == '__main__':
+    print(repr(example_grid.__name__))
+    print(repr(example_grid.__doc__))
```
Comment (PythonCHB) [inline students/guillaume/session02/grid.py:68]: why the repr?

File: students/guillaume/session02/series.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    for j in range(m):
+        print(repr(function_list[j].__name__))
+        print(repr(function_list[j].__doc__))
+        serie_lst = [function_list[j](i) for i in list_n]
+        serie_str = ', '.join(map(str, serie_lst))
+        print(serie_str)
+        print()
+
+    # Assert
+    assert fibonnaci(3) != fibonnaci(2) - fibonnaci(1), 'this is wrong'
```
Comment (PythonCHB) [inline students/guillaume/session02/series.py:53]: nice use of assert message.

very nice job on the whole thing.

File: students/guillaume/session03/rot13.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+
+from string import ascii_lowercase, ascii_uppercase
+
+def rot(n):
+	'''
+	Generate subsitution cypher table of value n
```
Comment (PythonCHB) [inline students/guillaume/session03/rot13.py:6]: very nice! I don't think I've seen this solution before.

File: students/guillaume/session03/rot13.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	Provide translation from encoded to natural
+	'''
+	return translation(string, 13, False)
+
+if __name__ == '__main__':
+	print(rot(13))
+	print(type(rot(13)))
+	print(rot13('testo'))
+	print(rot13_reverse('Zntargvp sebz bhgfvqr arne pbeare'))
+
```
Comment (PythonCHB) [inline students/guillaume/session03/rot13.py:42]: this is a very nice solution.

But do make sure you understand what ord() and chr() do as well.

File: students/guillaume/session03/slicing_lab.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    seq = (seq[-1:] + seq[1:l] + seq[0:1])
+    return seq
+
+
+def every_other_rem(seq):
+    '''
+    Print a sequence with every other item of the initial sequence
+    '''
+    l = len(seq)
+    seq_ret = seq[0:1]
```
Comment (PythonCHB) [inline students/guillaume/session03/slicing_lab.py:16]: you can do this with a step in the slice:

seq[::2]

File: students/guillaume/session03/slicing_lab.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        seq_ret = seq_ret + seq[i:i + 1]
+    return seq_ret
+
+
+def first_last_in_bet(seq):
+    '''
+    return a sequence with the first and last 4 items removed,
+    and every other item in between
+    '''
+    return every_other_rem(seq[4:len(seq) - 4])
```
Comment (PythonCHB) [inline students/guillaume/session03/slicing_lab.py:27]: you do this with the step in the slice right here.

File: students/guillaume/session03/slicing_lab.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def reverse_slic(seq):
+    '''
+    return a sequence reversed (just with slicing)
+    '''
+    l = len(seq)
+    seq_ret = seq[l - 1:l]
+    for i in range(l - 1, -1, -1):
```
Comment (PythonCHB) [inline students/guillaume/session03/slicing_lab.py:36]: this is a for loop -- not "just" slicing!

you did this in the print_grid exercise....

File: students/guillaume/session03/slicing_lab.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        third = l // 3
+        new_seq = seq[third:2 * third] + seq[2 * third: ] + seq[0:third]
+        return new_seq
+    print('len of the sequence is not a multiple of 3')
+    return None
+
+if __name__ == '__main__':
+
+    functions = [exch_first_last, every_other_rem, first_last_in_bet, reverse_slic, third]
+    test_lst = ['', 'tesu', 'ab',''.join(map(str,list(range(10)))), [1,2,3], list(range(24)), list(range(9)),['1third', '2third', '3third'], []]
```
Comment (PythonCHB) [inline students/guillaume/session03/slicing_lab.py:57]: nice!

**Review by PythonCHB — APPROVED** [review]
nice work!

**Pull request merged**